pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -371,5 +441,70 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
     for (int i = 0; i < nd; i++)
         nd_shape[i] = np_shape[i];
     return nd_shape;
+}
 
+/*
+** takes an array containing the shape of an array 'shape', number of a certain dimension 'nd', and the number of the array's dimensions
+** returns the stride (number of single elements to jump in a dimension to get to this dimension's next element) of the 'nd`th dimension
+*/
+
+int get_dimension_stride(int64_t *shape, int nd, int max_nd)
+{
+    int product = 1;
+
+    for (int i = nd; i < max_nd; ++i)
+        product *= shape[i];
+    return (product);
+}
+
+/*
+**  Takes an array needed to do the caculations
+**  An element number, representing an element's index if it were in a flattened (order_c/row major) array
+**  nd, representing the number of dimensions
+**  returns the element's index depending on its required memory layout (order_f/column major or order_c/row major)
+*/
+
+int element_index(t_ndarray arr, uint32_t element_num, int nd)
+{
+    if (arr.order == order_c)
+        return element_num;
+    if (nd == 0)
+        return (0);
+    if (nd == arr.nd)
+        return (element_num % arr.shape[nd - 1]) * arr.strides[nd - 1] + element_index(arr, element_num, nd - 1);
+    int true_index = (element_num / (get_dimension_stride(arr.shape, nd, arr.nd)));
+    if (true_index >= arr.shape[nd - 1])
+        true_index = true_index % arr.shape[nd - 1];
+    return (true_index * arr.strides[nd - 1] + element_index(arr, element_num, nd - 1));
+}
+
+bool is_same_shape(t_ndarray a, t_ndarray b)
+{
+    if (a.nd != b.nd)
+        return (false);
+    for (int i = 0; i < a.nd; ++i)
+    {
+        if (a.shape[i] != b.shape[i])
+            return (false);
+    }
+    return (true);
+}
+
+void array_copy_data(t_ndarray *dest, t_ndarray src, uint32_t offset)
+{
+    unsigned char *d = (unsigned char*)dest->raw_data;
+    unsigned char *s = (unsigned char*)src.raw_data;
+
+    if (!src.is_view && dest->order == src.order
+        && (src.order == order_c
+            || (src.order == order_f && is_same_shape(*dest, src))))
+        memcpy(d + offset * dest->type_size, s, src.buffer_size);
+    else
+    {
+        for (int32_t element_num = 0; element_num < src.length; ++element_num)
+        {
+            memcpy(d + ((element_index(*dest, element_num, dest->nd) + offset) * dest->type_size),
+                s + (element_index(src, element_num, src.nd) * src.type_size), src.type_size);

```

Comment:
I presume this is written like this as it was the only way to avoid a huge if/else block?

Reply:
Yes, so that we can copy straight to `raw_data`